pkg/flowaggregator/flowaggregator.go

```diff
@@ -724,3 +904,101 @@ func (fa *flowAggregator) createInfoElementForTemplateSet(ieName string, enterpr
 	}
 	return ie, nil
 }
+
+func (fa *flowAggregator) watchConfiguration(stopCh <-chan struct{}) {
+	klog.Info("Watching for FlowAggregator configuration file")
+	for {
+		select {
+		case <-stopCh:
+			return
+		case event, ok := <-fa.configWatcher.Events:
+			if err := fa.handleWatcherEvent(event, ok); err != nil {
+				// if the watcher cannot add mounted configMap or the configMap is not readable, we kill the flow-aggregator pod (serious error)
+				klog.Fatalf("Cannot watch or read configMap: %v", err)
+			}
+		case err, _ := <-fa.configWatcher.Errors:
+			if err != nil {
+				// if the error happens to watcher, we kill the flow-aggregator pod.
+				// watcher might be shut-down or broken in this situation.
+				klog.Fatalf("configWatcher err: %v", err)
+			}
+		}
+	}
+}
+
+func (fa *flowAggregator) handleWatcherEvent(event fsnotify.Event, ok bool) error {
+	if !ok {
+		return fmt.Errorf("configWatcher event channel closed")
+	}
+	// When mounted configmap is updating, watcher event "CHMOD" and "REMOVE" will happen.
+	// Need to add configFile back to watcher in order to watch the updated configMap.
+	if event.Op != fsnotify.Remove {
+		return nil
+	}
+	if err := fa.configWatcher.Add(fa.configFile); err != nil {
+		return fmt.Errorf("error when adding configFile back to watcher: %v", err)
+	}
+	data, err := ioutil.ReadFile(fa.configFile)
```

Comment:
`fa.configWatcher.Add` and `ReadFile` should succeed. When we update the volume configMap, the whole mounted configMap will be replaced by the new one with same name. Unless user manually modify the configMap's name or remove it, I think we are safe?

Reply:
I think watching the fir works fine. But there is an issue. The name of configMap is not showed in the event log for the first time we update the configMap. Not sure which event name should I pick? Since we also compare the data inside the file, I think as long as the event happens, we can update the changes based on that.

Here is the event.string()
Logs for the first time we update the volume configMap:
```
flowaggregator.go:916] "/etc/flow-aggregator/..2022_06_09_20_23_15.2653475576": CREATE
flowaggregator.go:916] "/etc/flow-aggregator/..2022_06_09_20_23_15.2653475576": CHMOD
flowaggregator.go:916] "/etc/flow-aggregator/..data_tmp": RENAME
flowaggregator.go:916] "/etc/flow-aggregator/..data": CREATE
flowaggregator.go:916] "/etc/flow-aggregator/..2022_06_09_20_23_14.3345969342": REMOVE
```
Logs after the first time:
```
flowaggregator.go:916] "/etc/flow-aggregator/..2022_06_09_20_26_51.2628093328": CREATE
flowaggregator.go:916] "/etc/flow-aggregator/..2022_06_09_20_26_51.2628093328": CHMOD
flowaggregator.go:916] "/etc/flow-aggregator/..data_tmp": RENAME
flowaggregator.go:916] "/etc/flow-aggregator/..data": CREATE
flowaggregator.go:916] "/etc/flow-aggregator/flow-aggregator.conf": CHMOD
flowaggregator.go:916] "/etc/flow-aggregator/flow-aggregator.conf": REMOVE
flowaggregator.go:916] "/etc/flow-aggregator/..2022_06_09_20_23_15.2653475576": REMOVE
```

Reply:
I am not sure how to interpret the events, but I assume that this is because `flow-aggregator.conf` is actually a symbolic link.
**If you don't think there is any issue with the original approach, you can stick to it.**
However, in the original approach, you assume that there is a `REMOVE` event for `/etc/flow-aggregator/flow-aggregator.conf`, so don't you have the same problem (event missing for first update)? Note that I don't see why the first update would be different from subsequent ones. Maybe it has something to do with how fsnotify follows symlinks.

Reply:
When I watched the conf file directly, these two events happened every time.
flowaggregator.go:916] "/etc/flow-aggregator/flow-aggregator.conf": CHMOD
flowaggregator.go:916] "/etc/flow-aggregator/flow-aggregator.conf": REMOVE

Reply:
I think it is better to watch the directory. Like you said, we don't need to add the file back to our watcher. In addition, when I was adding the unit test, I found that not every events is added to the event channel successfully. For example, in the unit test, I use os.chmod/delete/create sequentially and I expect there should be 2 events in the channel. However, I observed that:
1. 2 events (chmod, remove) in the channel every time in my local test.
2. 1 or 2 events in the channel when doing Unit test (ubuntu-latest)
3. only remove event in the channel when doing Unit test (windows-2019)

I modify the watcher to watch the directory, but I don't filter the event (we still check the bytes to decide whether we should move on). In this way, we don't add the file back whenever we update our fa Pod and we can avoid behavior difference when doing unit test on different OS.

Reply:
sounds good to me